IDs
===
For an entry to be added to the database, a unique ID must be used to
unambiguously retrieve that row.  Due to the structure of GFF and GTF files,
this must be handled differently depending on file type.

For GFF files, the "ID" attribute is used for the primary key if available,
otherwise the "Name" attribute if available, otherwise the "gene_name"
attribute if available, otherwise auto-generate a unique ID.

Autogenerated IDs are created as "featuretype:chrom:start:stop:strand".

GTF exons and CDSs typically do not have unique IDs associated with them; as a
result they will also use the auto-generated ID,
"featuretype:chrom:start:stop:strand".

In general, if the same genomic coordinates are specified for two features of
the same type but with different parents (say, a CDS is listed as two rows in
the GTF file, once for each isoform it is part of), there will only be a single
feature in the features table for that CDS but multiple entries in the
relations table, thus no information is lost.

In GTF files, genes and transcripts are not explicitly defined.  Transcripts
are inferred from the min/max coords of their child exons, while genes are
inferred from the min/max coords of their child transcripts.  In each case, the
inferred features are given the ID provided in the GTF file in the "gene_id"
and "transcript_id" attributes of their child exons.
