Attributes are either a list (if a comma is in the field) or a string.

Each feature needs a unique ID (hereafter called "uid" to differentiate
from the "ID" attribute) so that it can be placed in the database.  For GFF
files, the uid is the "ID" field if available, "Name" field as a fallback,
or autogenerated as a last resort.  This means that IDs cannot be
duplicated in a GFF file (this is typically not a problem).

For GTF files, it gets a little more complex.  GTF files only contain
exons/cds/start and stop codons.  Each of these features typically does not
have an ID that can be used for a uid.  

As a result, GTF exons and CDSs always have autogenerated uids created from
"featuretype:chrom:start:stop:strand".  This information should hopefully be
enough to uniquely identify the exon.  If the same genomic coordinates are
specified for two features of the same type but with different parents,
there will only be a single feature in the features table but multiple
entries in the relations table.

Genes and transcripts are constructed from the component exon information
in a GTF file.  These genes and transcripts have valid uids taken from the
"gene_id" and "transcript_id" attributes of their children exons.
